### PR TITLE
Added Optimistic Concurrency Pattern

### LINKED ISSUE #1280: Optimistic Concurrency pattern
http://ommolketab.ir/aaf-lib/axkwht7wxrhvgs2aqkxse8hihyu9zv.pdf chapter 10.2

### PR DESCRIPTION
**Pull request by JeffersonJefferson-pixel**
Pull request title

- Added Optimistic Concurrency Pattern
- fixes #1280

Pull request description

- Implements Optimistic Concurrency Pattern based on provided resource: http://ommolketab.ir/aaf-lib/axkwht7wxrhvgs2aqkxse8hihyu9zv.pdf chapter 10.2
- Also took inspiration from https://www.javacodegeeks.com/2012/11/jpahibernate-version-based-optimistic-concurrency-control.html

### CODE REVIEW

File: optimistic-concurrency/pom.xml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    </artifactId>
+
+    <properties>
+        <maven.compiler.source>
+            11
+        </maven.compiler.source>
+        <maven.compiler.target>
+            11
+        </maven.compiler.target>
+    </properties>
```
Comment (iluwatar) [inline optimistic-concurrency/pom.xml:32]: Leave out, use the versions that the parent pom.xml specifies

File: optimistic-concurrency/pom.xml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.hibernate.javax.persistence</groupId>
+            <artifactId>hibernate-jpa-2.1-api</artifactId>
+            <version>1.0.2.Final</version>
```
Comment (iluwatar) [inline optimistic-concurrency/pom.xml:38]: Use the versions from the parent pom.xml

File: optimistic-concurrency/pom.xml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+            <version>1.4.200</version>
+        </dependency>
+    </dependencies>
```
Comment (iluwatar) [inline optimistic-concurrency/pom.xml:80]: Add maven-assembly-plugin so we can execute this example from the command line

File: optimistic-concurrency/src/main/java/com/iluwatar/optimisticconcurrency/App.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import javax.persistence.Persistence;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Optimistic Concurrency pattern.
+ * Increment version during every update.
+ * Attempts to update object using old version fail
+ */
```
Comment (iluwatar) [inline optimistic-concurrency/src/main/java/com/iluwatar/optimisticconcurrency/App.java:14]: Please add more verbose explanation what the pattern is and how this example implements it

File: optimistic-concurrency/src/main/java/com/iluwatar/optimisticconcurrency/App.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          LOGGER.info("There are {} apples left.",
+              appleRes.getAmountInStock());
+        }
+      }
+
+      emf.close();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
```
Comment (iluwatar) [inline optimistic-concurrency/src/main/java/com/iluwatar/optimisticconcurrency/App.java:101]: It's very hard to see what is going on in this example without explanation. Please add comments.

**Review by iluwatar — CHANGES_REQUESTED** [review]

**Comment by mohanaraosv** [comment]
@iluwatar Shall I continue this and give credit to the original Author?

**Comment by iluwatar** [comment]
Oh yes, that would be great @mohanaraosv Please comment on the issue and I'll assign it for you.

**Pull request closed without merging**